network/wifi/at_parser.c

```diff
@@ -204,12 +184,41 @@ struct at_desc {
 	}			ipd_stat;
 	/* Variable where current message length is built */
 	uint32_t		ipd_len;
+	/* Will be called when a new connection is created or closed */
+	void			(*connection_callback)(void *ctx, enum at_event,
+			uint32_t conn_id, struct circular_buffer **cb);
+	/* Context that will be passed to the callback */
+	void			*callback_ctx;
 };
 
 /******************************************************************************/
 /************************ Functions Definitions *******************************/
 /******************************************************************************/
 
+static void dispatch_event(struct at_desc *desc, enum at_event event)
+{
+	struct connection_desc *conn;
+
+	conn = &desc->conn[desc->current_conn];
+
+	if (event == AT_NEW_CONNECTION) {
+		if (!conn->active) {
+			desc->connection_callback(desc->callback_ctx, event,
```

Comment:
i think the callback can be moved outside of this and called only once.

Reply:
I can't, this is the calling flow. If I change that will be hard to understand.
For example AT_NEW_CONNECTION is generated multiple times for the same connection and connection_callback should be called only the first time. But,  connection_callback from AT_CLOSED_CONNECTION event should be called every time.

Reply:
```
if (event == AT_NEW_CONNECTION) {
			if (conn->cbuff)
				conn->active = true;
			//else todo close connection
		}
} else if (event == AT_CLOSED_CONNECTION) {
		conn->active = false;
		conn->cbuff = NULL;
		
}
if (!conn->active) {
desc->connection_callback(desc->callback_ctx, event,
			   desc->current_conn,
			  &conn->cbuff);

Reply:
```
if (conn->cbuff)
 conn->active = true
```
should be done after  connection_callback
also,
when `event == AT_CLOSED_CONNECTION` it is better if the last parameter is NULL

Reply:
Well if the return code for the callback is not checked then it does not matter where you change conn->active.
the last parameter is NULL when you call the connection_callback because it is changed inside the if statement with:
```
 conn->cbuff = NULL;

Reply:
conn->cbuff can be set inside connection_callback. To the callback is passed **&**conn->cbuff

Reply:
Where is connection_calback to be able to see what it does exactly?

Reply:
I splitted dispach_event: half in start_conn_read and half in check_conn_id